// appium/build/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppiumModuleRoot = exports.inspect = exports.npmPackage = exports.V6_BROADCAST_IP = exports.V4_BROADCAST_IP = void 0;
exports.makeNonW3cCapsError = makeNonW3cCapsError;
exports.parseCapsForInnerDriver = parseCapsForInnerDriver;
exports.insertAppiumPrefixes = insertAppiumPrefixes;
exports.removeAppiumPrefixes = removeAppiumPrefixes;
exports.adjustNodePath = adjustNodePath;
exports.pullSettings = pullSettings;
exports.isServerCommandArgs = isServerCommandArgs;
exports.isSetupCommandArgs = isSetupCommandArgs;
exports.isExtensionCommandArgs = isExtensionCommandArgs;
exports.isDriverCommandArgs = isDriverCommandArgs;
exports.isPluginCommandArgs = isPluginCommandArgs;
exports.fetchInterfaces = fetchInterfaces;
exports.adler32 = adler32;
exports.isBroadcastIp = isBroadcastIp;
const lodash_1 = __importDefault(require("lodash"));
const logger_1 = __importDefault(require("./logger"));
const base_driver_1 = require("@appium/base-driver");
const node_util_1 = require("node:util");
const support_1 = require("@appium/support");
const node_path_1 = __importDefault(require("node:path"));
const constants_1 = require("./constants");
const node_os_1 = __importDefault(require("node:os"));
const W3C_APPIUM_PREFIX = 'appium';
const STANDARD_CAPS_LOWERCASE = new Set([...base_driver_1.STANDARD_CAPS].map((cap) => cap.toLowerCase()));
exports.V4_BROADCAST_IP = '0.0.0.0';
exports.V6_BROADCAST_IP = '::';
exports.npmPackage = support_1.fs.readPackageJsonFrom(__dirname);
/** If stdout is a TTY; used for tighter control over log output. */
const isStdoutTTY = process.stdout.isTTY;
/**
 * Creates an error when a session receives non-W3C capabilities.
 */
function makeNonW3cCapsError() {
    return new base_driver_1.errors.SessionNotCreatedError('Session capabilities format must comply to the W3C standard. Make sure your client is up to date. ' +
        'See https://www.w3.org/TR/webdriver/#new-session for more details.');
}
/**
 * Logs a value to the console using the info logger (with util.inspect formatting).
 */
exports.inspect = lodash_1.default.flow(lodash_1.default.partialRight(node_util_1.inspect, {
    colors: true,
    depth: null,
    compact: !isStdoutTTY,
}), (...args) => {
    logger_1.default.info(...args);
});
/**
 * Parses W3C capabilities for the inner driver and applies defaults.
 *
 * @returns Parsed caps or an invalid result with an error.
 */
function parseCapsForInnerDriver(w3cCapabilities, constraints = {}, defaultCapabilities = {}) {
    if (!(0, base_driver_1.isW3cCaps)(w3cCapabilities)) {
        return { error: makeNonW3cCapsError() };
    }
    let desiredCaps = {};
    // eslint-disable-next-line prefer-const -- assigned in success path after try
    let processedW3CCapabilities;
    w3cCapabilities = lodash_1.default.cloneDeep(w3cCapabilities);
    defaultCapabilities = lodash_1.default.cloneDeep(defaultCapabilities);
    if (!lodash_1.default.isEmpty(defaultCapabilities)) {
        for (const [defaultCapKey, defaultCapValue] of lodash_1.default.toPairs(defaultCapabilities)) {
            let isCapAlreadySet = false;
            for (const firstMatchEntry of w3cCapabilities.firstMatch ?? []) {
                if (lodash_1.default.isPlainObject(firstMatchEntry) &&
                    lodash_1.default.has(removeAppiumPrefixes(firstMatchEntry), removeAppiumPrefix(defaultCapKey))) {
                    isCapAlreadySet = true;
                    break;
                }
            }
            isCapAlreadySet =
                isCapAlreadySet ||
                    (lodash_1.default.isPlainObject(w3cCapabilities.alwaysMatch) &&
                        lodash_1.default.has(removeAppiumPrefixes(w3cCapabilities.alwaysMatch), removeAppiumPrefix(defaultCapKey)));
            if (isCapAlreadySet) {
                continue;
            }
            if (lodash_1.default.isEmpty(w3cCapabilities.firstMatch)) {
                w3cCapabilities.firstMatch = [{ [defaultCapKey]: defaultCapValue }];
            }
            else {
                w3cCapabilities.firstMatch[0][defaultCapKey] = defaultCapValue;
            }
        }
    }
    try {
        desiredCaps = (0, base_driver_1.processCapabilities)(w3cCapabilities, constraints, true);
    }
    catch (error) {
        logger_1.default.info(`Could not parse W3C capabilities: ${error.message}`);
        return {
            desiredCaps,
            processedW3CCapabilities,
            error: error,
        };
    }
    processedW3CCapabilities = {
        alwaysMatch: { ...insertAppiumPrefixes(desiredCaps) },
        firstMatch: [{}],
    };
    return {
        desiredCaps,
        processedW3CCapabilities,
    };
}
/**
 * Prefixes capability keys with `appium:` where appropriate.
 */
function insertAppiumPrefixes(caps) {
    return lodash_1.default.mapKeys(caps, (_, key) => STANDARD_CAPS_LOWERCASE.has(key.toLowerCase()) || key.includes(':')
        ? key
        : `${W3C_APPIUM_PREFIX}:${key}`);
}
/**
 * Removes `appium:` prefix from capability keys.
 */
function removeAppiumPrefixes(caps) {
    return lodash_1.default.mapKeys(caps, (_, key) => removeAppiumPrefix(key));
}
/**
 * Returns the root directory of the Appium module (memoized).
 *
 * @throws {Error} If the appium module root cannot be determined.
 */
exports.getAppiumModuleRoot = lodash_1.default.memoize(function getAppiumModuleRoot() {
    const selfRoot = support_1.node.getModuleRootSync('appium', __filename);
    if (!selfRoot) {
        throw new Error('Cannot find the appium module root. This is likely a bug in Appium.');
    }
    return selfRoot;
});
/**
 * Adjusts NODE_PATH so CJS drivers/plugins can load peer deps. Does not work with ESM.
 */
function adjustNodePath() {
    let appiumModuleSearchRoot;
    try {
        appiumModuleSearchRoot = node_path_1.default.dirname((0, exports.getAppiumModuleRoot)());
    }
    catch (error) {
        logger_1.default.warn(error.message);
        return;
    }
    const refreshRequirePaths = () => {
        try {
            // Private API; see https://gist.github.com/branneman/8048520#7-the-hack
            require('node:module').Module._initPaths();
            return true;
        }
        catch {
            return false;
        }
    };
    if (!process.env.NODE_PATH) {
        process.env.NODE_PATH = appiumModuleSearchRoot;
        if (refreshRequirePaths()) {
            process.env.APPIUM_OMIT_PEER_DEPS = '1';
        }
        else {
            delete process.env.NODE_PATH;
        }
        return;
    }
    const nodePathParts = process.env.NODE_PATH.split(node_path_1.default.delimiter);
    if (nodePathParts.includes(appiumModuleSearchRoot)) {
        process.env.APPIUM_OMIT_PEER_DEPS = '1';
        return;
    }
    nodePathParts.push(appiumModuleSearchRoot);
    process.env.NODE_PATH = nodePathParts.join(node_path_1.default.delimiter);
    if (refreshRequirePaths()) {
        process.env.APPIUM_OMIT_PEER_DEPS = '1';
    }
    else {
        process.env.NODE_PATH = lodash_1.default.without(nodePathParts, appiumModuleSearchRoot).join(node_path_1.default.delimiter);
    }
}
/**
 * Pulls Appium settings from capabilities (mutates caps). Supports
 * `settings[key]: value` and `settings: { key: value }`.
 *
 * @returns Parsed settings object; empty if none found.
 */
function pullSettings(caps) {
    if (!lodash_1.default.isPlainObject(caps) || lodash_1.default.isEmpty(caps)) {
        return {};
    }
    const result = {};
    const singleSettings = {};
    for (const [key, value] of lodash_1.default.toPairs(caps)) {
        let match;
        if (/^(s|appium:s)ettings$/.test(key) && lodash_1.default.isPlainObject(value)) {
            Object.assign(result, value);
            delete caps[key];
        }
        else if ((match = /^(s|appium:s)ettings\[(\S+)\]$/.exec(key))) {
            singleSettings[match[2]] = value;
            delete caps[key];
        }
    }
    if (!lodash_1.default.isEmpty(singleSettings)) {
        Object.assign(result, singleSettings);
    }
    return result;
}
/**
 * Type guard: args are for the server command.
 */
function isServerCommandArgs(args) {
    return args.subcommand === constants_1.SERVER_SUBCOMMAND;
}
/**
 * Type guard: args are for the setup command.
 */
function isSetupCommandArgs(args) {
    return args.subcommand === constants_1.SETUP_SUBCOMMAND;
}
/**
 * Type guard: args are for an extension command (driver or plugin).
 */
function isExtensionCommandArgs(args) {
    return args.subcommand === constants_1.DRIVER_TYPE || args.subcommand === constants_1.PLUGIN_TYPE;
}
/**
 * Type guard: args are for a driver extension command.
 */
function isDriverCommandArgs(args) {
    return args.subcommand === constants_1.DRIVER_TYPE;
}
/**
 * Type guard: args are for a plugin extension command.
 */
function isPluginCommandArgs(args) {
    return args.subcommand === constants_1.PLUGIN_TYPE;
}
/**
 * Returns network interfaces for the given IP family.
 *
 * @param family - 4 for IPv4, 6 for IPv6, or null for all.
 */
function fetchInterfaces(family = null) {
    let familyValue = null;
    if (family === 4) {
        familyValue = [4, 'IPv4'];
    }
    else if (family === 6) {
        familyValue = [6, 'IPv6'];
    }
    const ifaces = lodash_1.default.values(node_os_1.default.networkInterfaces()).filter(Boolean);
    return lodash_1.default.flatMap(ifaces).filter((info) => !familyValue || familyValue.includes(info.family));
}
/**
 * Adler-32 checksum (see https://github.com/SheetJS/js-adler32).
 */
function adler32(str, seed = null) {
    let a = 1, b = 0, M = 0, c = 0, d = 0;
    const L = str.length;
    if (typeof seed === 'number') {
        a = seed & 0xffff;
        b = seed >>> 16;
    }
    for (let i = 0; i < L;) {
        M = Math.min(L - i, 2918);
        while (M > 0) {
            c = str.charCodeAt(i++);
            if (c < 0x80) {
                a += c;
            }
            else if (c < 0x800) {
                a += 192 | ((c >> 6) & 31);
                b += a;
                --M;
                a += 128 | (c & 63);
            }
            else if (c >= 0xd800 && c < 0xe000) {
                c = (c & 1023) + 64;
                d = str.charCodeAt(i++) & 1023;
                a += 240 | ((c >> 8) & 7);
                b += a;
                --M;
                a += 128 | ((c >> 2) & 63);
                b += a;
                --M;
                a += 128 | ((d >> 6) & 15) | ((c & 3) << 4);
                b += a;
                --M;
                a += 128 | (d & 63);
            }
            else {
                a += 224 | ((c >> 12) & 15);
                b += a;
                --M;
                a += 128 | ((c >> 6) & 63);
                b += a;
                --M;
                a += 128 | (c & 63);
            }
            b += a;
            --M;
        }
        a = 15 * (a >>> 16) + (a & 65535);
        b = 15 * (b >>> 16) + (b & 65535);
    }
    return ((b % 65521) << 16) | (a % 65521);
}
/**
 * Returns true if the address is a broadcast IP (0.0.0.0 or ::).
 */
function isBroadcastIp(address) {
    return [exports.V4_BROADCAST_IP, exports.V6_BROADCAST_IP, `[${exports.V6_BROADCAST_IP}]`].includes(address);
}
// #region private helpers
function removeAppiumPrefix(key) {
    const prefix = `${W3C_APPIUM_PREFIX}:`;
    return lodash_1.default.startsWith(key, prefix) ? key.substring(prefix.length) : key;
}
// #endregion
